mlir/lib/Dialect/UB/IR/UBOps.cpp

```diff
@@ -57,8 +58,33 @@ Operation *UBDialect::materializeConstant(OpBuilder &builder, Attribute value,
   return nullptr;
 }
 
+//===----------------------------------------------------------------------===//
+// PoisonOp
+//===----------------------------------------------------------------------===//
+
 OpFoldResult PoisonOp::fold(FoldAdaptor /*adaptor*/) { return getValue(); }
 
+//===----------------------------------------------------------------------===//
+// UnreachableOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult UnreachableOp::canonicalize(UnreachableOp unreachableOp,
+                                          PatternRewriter &rewriter) {
+  Block *block = unreachableOp->getBlock();
+  if (llvm::hasSingleElement(*block))
+    return rewriter.notifyMatchFailure(
+        unreachableOp, "unreachable op is the only operation in the block");
+
+  // Erase all other operations in the block. They must be dead.
+  for (Operation &op : llvm::make_early_inc_range(*block)) {
+    if (&op == unreachableOp.getOperation())
+      continue;
+    op.dropAllUses();
+    rewriter.eraseOp(&op);
+  }
```

Comment:
We should TODO that:

- this assumes we don't have calls that are "no return".
- this assumes that loops terminates.
- this assumes that nothing interrupts the control-flow (which is fine until early-exit is added).

Reply:
Maybe we should have a trait "AlwaysForwardProgress" on operations to allow this kind of transformations?

Reply:
I didn't think of infinite loops. I am wondering if this canonicalization is actually safe. I'm just merging the `cf.cond_br` canonicalization for now, so we can discuss further. I feel like this kind of "canonicalization" may be better suited for a separate pass. (Something like `-remove-dead-values` maybe?) We could also make this optimization customizable (turn off/on via pass option).